projects/plugins/jetpack/extensions/store/membership-products/resolvers.js

```diff
@@ -25,84 +25,102 @@ import { getMessageByProductType } from '../../shared/components/product-managem
 import executionLock from '../../shared/execution-lock';
 
 const EXECUTION_KEY = 'membership-products-resolver-getProducts';
+let hydratedFromAPI = false;
 
-export const getProducts = (
-	productType = PRODUCT_TYPE_PAYMENT_PLAN,
-	selectedProductId = 0,
-	setSelectedProductId = () => {}
-) => async ( { dispatch, registry } ) => {
-	await executionLock.blockExecution( EXECUTION_KEY );
-	const lock = executionLock.acquire( EXECUTION_KEY );
-
+const fetchMemberships = async () => {
 	const origin = getQueryArg( window.location.href, 'origin' );
 	const path = addQueryArgs( '/wpcom/v2/memberships/status', {
 		source: origin === 'https://wordpress.com' ? 'gutenberg-wpcom' : 'gutenberg',
 		type: 'all',
 		is_editable: true,
 	} );
 
-	try {
-		const response = await apiFetch( { path, method: 'GET' } );
-		if ( ! response && typeof response !== 'object' ) {
-			return;
-		}
+	const response = await apiFetch( { path, method: 'GET' } );
 
-		/**
-		 * WP_Error returns a list of errors with custom names:
-		 * `errors: { foo: [ 'message' ], bar: [ 'message' ] }`
-		 * Since we don't know their names, to get the message, we transform the object
-		 * into an array, and just pick the first message of the first error.
-		 *
-		 * @see https://developer.wordpress.org/reference/classes/wp_error/
-		 */
-		const wpError = response?.errors && Object.values( response.errors )?.[ 0 ]?.[ 0 ];
-		if ( wpError ) {
-			dispatch( setApiState( API_STATE_NOTCONNECTED ) );
-			onError( wpError, registry );
-			return;
-		}
+	if ( ! response && typeof response !== 'object' ) {
+		throw new Error( 'Unexpected API response' );
+	}
+
+	/**
+	 * WP_Error returns a list of errors with custom names:
+	 * `errors: { foo: [ 'message' ], bar: [ 'message' ] }`
+	 * Since we don't know their names, to get the message, we transform the object
+	 * into an array, and just pick the first message of the first error.
+	 *
+	 * @see https://developer.wordpress.org/reference/classes/wp_error/
+	 */
+	const wpError = response?.errors && Object.values( response.errors )?.[ 0 ]?.[ 0 ];
+	if ( wpError ) {
+		throw new Error( wpError );
+	}
 
-		const postId = registry.select( editorStore ).getCurrentPostId();
+	return response;
+};
+
+const mapAPIResponseToMembershipProductsStoreData = ( response, registry, dispatch ) => {
+	const postId = registry.select( editorStore ).getCurrentPostId();
 
-		dispatch( setConnectUrl( getConnectUrl( postId, response.connect_url ) ) );
-		dispatch( setShouldUpgrade( response.should_upgrade_to_access_memberships ) );
-		dispatch( setSiteSlug( response.site_slug ) );
-		dispatch( setUpgradeUrl( response.upgrade_url ) );
+	dispatch( setConnectUrl( getConnectUrl( postId, response.connect_url ) ) );
+	dispatch( setShouldUpgrade( response.should_upgrade_to_access_memberships ) );
+	dispatch( setSiteSlug( response.site_slug ) );
+	dispatch( setUpgradeUrl( response.upgrade_url ) );
+	dispatch( setProducts( response.products ) );
```

Comment:
Now we always load the products that we get from the API since this is the initial loading of the store and is not conditional.